rule-audit-client/src/redux/modules/searchMatches/__tests__/thunks.spec.ts

```diff
@@ -0,0 +1,249 @@
+    // We then call Typerighter with those articles, and Typerighter
+    // returns matches for the last two articles.
+    expect(mockCapiService).toBeCalledTimes(2);
+    expect(mockTyperighterService).toBeCalledTimes(4);
+
+    capiResponse.results.forEach((article) => {
+      const articleInState = capiSelectors.selectById(state, article.id);
+      expect(articleInState).toMatchObject(article);
+    });
+
+    expect(selectors.selectSearchMatchesArticleIds(state, false)).toEqual(
+      secondCapiResponse.results.map(_ => _.id)
+    );
+  });
+  it("should stop asking for articles if the cancel action is dispatched", async () => {
+    const store = createStore();
+    store.dispatch(actions.doSetSearchMatchesLimit(2));
+
+    mockResponses(
+      [capiResponse],
+      [
+        typerighterResponse,
+        emptyTyperighterResponse
+      ]
+    );
+
+   const thunk = (store.dispatch as ThunkDispatch<
+      AppTypes.RootState,
+      {},
+      AppTypes.RootAction
+    >)(
+      doSearchMatches(
+        "query",
+        [],
+        [],
+        mockCapiService,
+        mockTyperighterService
+      )
+    );
+
+    store.dispatch(actions.doSearchMatchesEnd());
+    await thunk;
+
+    const state = store.getState();
+
+    // We call CAPI, and it's returned two articles.
+    // We then call Typerighter with those articles, and Typerighter
+    // returns matches for one. We call CAPI again, but the user
+    // cancels the search before the second CAPI request is processed,
+    // so nothing else should go to Typerighter.
+    expect(mockCapiService).toBeCalledTimes(1);
+    expect(mockTyperighterService).toBeCalledTimes(2);
+
+    capiResponse.results.forEach((article) => {
+      const articleInState = capiSelectors.selectById(state, article.id);
+      expect(articleInState).toMatchObject(article);
+    });
+
+    expect(selectors.selectSearchMatchesArticleIds(state, false)).toEqual([capiResponse.results[0].id]);
+  });
+  it("should stop asking for articles if no more remain", async () => {
+    const store = createStore();
+    store.dispatch(actions.doSetSearchMatchesLimit(2));
+
+    mockResponses(
+      [createCapiResponse(2, 0, 1)],
+      [
+        emptyTyperighterResponse,
+        emptyTyperighterResponse
+      ]
+    );
+
+   const thunk = (store.dispatch as ThunkDispatch<
+      AppTypes.RootState,
+      {},
+      AppTypes.RootAction
+    >)(
+      doSearchMatches(
+        "query",
+        [],
+        [],
+        mockCapiService,
+        mockTyperighterService
+      )
+    );
+
+    await thunk;
+
+    const state = store.getState();
+
+    // We call CAPI, and it's returned two articles.
+    // We then call Typerighter with those articles, and Typerighter
+    // returns matches for one. We call CAPI again, but the user
+    // cancels the search before the second CAPI request is processed,
+    // so nothing else should go to Typerighter.
```

Comment:
Really appreciate the thorough explanation of these different scenarios 👍

Reply:
Oh, that's great! I found it difficult to express the scenarios as code – glad it was helpful.

Reply:
The only problem is that it's easy to write things that are out of sync with the code! I've just spotted this below – I'll amend.